step 2 PR

ilbender opened the pull request:
the function feedLstToDbLst is a bit redundant but it makes line 105 in app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt a bit more readable - should i keep it as is or inline the function or simply use the lambda there?

ilbender committed:
implemented the next step (i.e saving the data to a persistent database after fetching it from a remote source)
simplified the conversion of lists using lambda expressions.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/database/Entities.kt:
```diff
@@ -1,5 +1,13 @@
+@Entity
+data class UserProject(
+    @PrimaryKey val id: String,
+    @ColumnInfo(name = "thumbnail_Url") val thumbnailUrl : String,
```
The convention for the snake case is to use all lowercase (`thumbnail_url`).
This is for readability. SQLite is supposed to ignore the case.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/database/DataAccessObjects.kt:
```diff
@@ -1,5 +1,36 @@
+    @Query("SELECT * FROM userproject WHERE id IN (:userIds)")
+    fun getAllByIds(userIds : List<String>) : Observable<List<UserProject>>
+
+    @Query("SELECT * FROM userproject WHERE id LIKE :uid")
```
`LIKE` is used for patterns, and for IDs we want an exact match.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/database/DataAccessObjects.kt:
```diff
@@ -1,5 +1,36 @@
+
+    @Query("DELETE FROM userproject")
+    fun deleteAll() : Completable
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
```
👍

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/database/DataAccessObjects.kt:
```diff
@@ -1,5 +1,36 @@
+    fun delete(userProject: UserProject)
+
+    @Query("SELECT COUNT(*) FROM userproject")
+    fun getUserCount() : Single<Int>
```
`getProjectCount` is probably more correct.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/network/FeedApiResponseGenerator.kt:
```diff
@@ -19,7 +19,7 @@ object FeedApiResponseGenerator {
         feedApiService = retrofit.create(FeedApiService::class.java)
     }
-    fun createFeedApiService() : FeedApiService {
+    fun getFeedApiService() : FeedApiService {
```
You can make `feedApiService` public and then you don't need a getter:
`val feedApiService = retrofit.create(FeedApiService::class.java)`

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -34,30 +36,28 @@ open class FeedViewModel () : ViewModel() {
-            val currItem = FeedItem(item.id, responseUrlToThumbnailUrl(item) ,item.isPremium)
-            feedItemLst.add(currItem)
-        }
+        val feedItemLst : List<FeedItem> = rawResponseList.map {
```
This is OK.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -97,23 +97,38 @@ open class FeedViewModel () : ViewModel() {
     }
+
+    private fun saveToDb(){
+        val DB_NAME = "user_project_db.db" //TODO: verify
```
It's OK.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -97,23 +97,38 @@ open class FeedViewModel () : ViewModel() {
+
+    private fun saveToDb(){
+        val DB_NAME = "user_project_db.db" //TODO: verify
+        val db = Room.databaseBuilder(getApplication(),FeedDatabase::class.java,DB_NAME).build()
```
Android Studio can format the code for you when you are feeling lazy :)
Option+Shift+Command+L

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -97,23 +97,38 @@ open class FeedViewModel () : ViewModel() {
+        val db = Room.databaseBuilder(getApplication(),FeedDatabase::class.java,DB_NAME).build()
+        val userProjectDao = db.userProjectDao()
+        val disposable = userProjectDao.insertAll(feedLstToDbLst(stateInternal.value?.feedItems!!))
+            .subscribeOn(Schedulers.io()).subscribe({},{ error -> handleNetworkError(error)}) //TODO inline the function call or keep as is for readability?
```
Longs lines are harder to read. Let's break this one at the dot.

bmax-ltx replied:
This is not a network error. 
Let's rename `handleNetworkError` to something more general.

bmax-ltx replied:
I would put 
`val dbList = feedLstToDbLst(stateInternal.value?.feedItems!!) `
to a separate line.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -97,23 +97,38 @@ open class FeedViewModel () : ViewModel() {
\ No newline at end of file
+}
+
+fun feedLstToDbLst(feedItemLst : List<FeedItem>) : List<UserProject>{ //TODO remove? add inline?
```
It's OK to have conversion functions like this.

bmax-ltx reviewed (changes requested):
Nice!

ilbender committed:
1. changed FeedApiResponseGenerator so that theres no need for a getter.
2.renamed few methods in FeedViewModel
3.renamed columns as per snake case conventions
4. renamed getUserCount in DataAccessObjects.kt to getUserProjectCount
5. changed getById query

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -90,12 +94,25 @@ open class FeedViewModel () : ViewModel() {
+
+    private fun saveToDb() {
+        val DB_NAME = "user_project_db.db"
+        val db = Room.databaseBuilder(getApplication(), FeedDatabase::class.java, DB_NAME).build()
```
Let's move building the database outside of the function.

bmax-ltx reviewed (approved):
Good!
